Add GaussianQuadrature2D

Mateasek opened the pull request:
Adds `GaussianQuadrature2D` class mentioned in #475.

Adds also integrators (1D, 2D) to our documentation.

This PR will be marked as draft until PR #473 is merged.

Mateasek committed:
Add 2D Gauss-Legendre quadrature integrator

Mateasek committed:
Add tests for GaussianQuadrature2D

Mateasek committed:
Add Integrator2D documentation

Mateasek committed:
Add CHANGELOG.md record

Mateasek committed:
Merge branch 'development' into feature/GaussianQuadrature2D

Mateasek committed:
Merge branch 'development' into feature/GaussianQuadrature2D

jacklovell commented on cherab/core/math/integrators/__init__.pxd:
```diff
@@ -17,5 +17,5 @@
 # under the Licence.
 
 from cherab.core.math.integrators.integrators1d cimport Integrator1D, GaussianQuadrature
```
For consistency, we should rename `GaussianQuadrature` to `GaussianQuadrature1D` to distinguish it from the new 2D class. But I suggest keeping `GaussianQuadrature` as an alias to maintain backwards compatibility for now, and deprecate it in another release or so (removed in Cherab 2.0).

Mateasek replied:
I Agree. I added `GaussianQuadrature1D` and marked `GaussianQuadrature` as deprecated.

jacklovell commented on cherab/core/math/integrators/__init__.py:
```diff
@@ -17,4 +17,4 @@
 # under the Licence.
 
 from .integrators1d import Integrator1D, GaussianQuadrature
```
Ditto `GaussianQuadrature1D`

Mateasek replied:
done

jacklovell commented on cherab/core/math/integrators/integrators2d.pxd:
```diff
@@ -16,6 +16,8 @@
 # See the Licence for the specific language governing permissions and limitations
 # under the Licence.
 
+from numpy cimport ndarray
```
This is unnecessary: you don't make use of the typed `_x_roots`, `_x_weights`, `_y_roots` and `_y_weights` in any tight loops: they're only assigned to per the output of `roots_legendre`, which goes through the Python layer anyway. So you can cdef these as `object` instead of `ndarray` and save a numpy `cimport`, which will speed up compilation.

All the typed access is (correctly) done through the memoryviews ayway.

Mateasek replied:
true, removed

jacklovell commented on cherab/core/math/integrators/integrators2d.pyx:
```diff
@@ -59,3 +65,305 @@ cdef class Integrator2D:
+        f\\left( B \\xi_i + A, \\, D(x) \\eta_j + C(x) \\right) \\,
+        B \\cdot D(x)
+    where:
+        -  :math:`x_{\mathrm{lower}}`: Lower limit of integration for the x-dimension.
+        -  :math:`x_{\mathrm{upper}}`: Upper limit of integration for the x-dimension.
+        -  :math:`y_{\mathrm{lower}}(x)`: Lower limit of integration for the y-dimension, a function of :math:`x`.
+        -  :math:`y_{\mathrm{upper}}(x)`: Upper limit of integration for the y-dimension, a function of :math:`x`.
```
Either double-escape `\mathrm` or (preferably) use a raw string for the docstring (`r'''`) and go with single backslashes for all the Latex code.

Mateasek replied:
made the docstring a raw string

jacklovell replied:
You'll also need to un-escape the backslashes for TeX macros elsewhere in the docstring now that it is a raw string: a single backslash is needed everywhere otherwise the maths isn't rendered correctly when building the HTML docs.

There should also be a blank line after the `.. math::` block and before `where:`, otherwise Sphinx complains with `WARNING: Explicit markup ends without a blank line; unexpected unindent. [docutils]`

`./dev/build_docs.sh` from the root directory of the repo is good for locally checking how the docstrings render. You'll need the `sphinx`, `sphinx-tabs` and `sphinx-rtd-theme` Python packages installed.

Mateasek replied:
Thanks, I hope that now it is ok.

jacklovell commented on cherab/core/math/integrators/integrators2d.pyx:
```diff
@@ -59,3 +65,305 @@ cdef class Integrator2D:
+
+        self._rtol = value
+
+    cdef _build_cache(self):
```
This doesn't need to be `cdef`, a simple `def` will do.

Mateasek replied:
Again, here I went with what was done for `GaussianQuadrature1D`. I have no preference here, but I would prefer having it the same way for both 1D and 2D cases. Do you prefer me to change it to a python method in both 1D and 2D classes @jacklovell ?

jacklovell replied:
OK, fair enough. For consistency with what's already there `cdef` is fine.

jacklovell commented on cherab/core/math/integrators/integrators2d.pyx:
```diff
@@ -59,3 +65,305 @@ cdef class Integrator2D:
+    :param int x_max_order: Maximum order on Gaussian quadrature in the x dimension. Default is `50`.
+    :param int x_min_order: Minimum order on Gaussian quadrature in the x dimension. Default is `1`.
+    :param int y_max_order: Maximum order on Gaussian quadrature in the y dimension. Default is `50`.
+    :param int y_min_order: Minimum order on Gaussian quadrature in the y dimension. Default is `1`.
```
I think it's worth clarifying here that `order` is iteratively increased from `min_order` to `max_order` until the desired tolerance is reached, and if either x or y reaches max order before tolerance is reached the other one continues to increase order while the first is pegged at its max order. It takes a lot of reading of the code to realise this.

This explanation is also missing from the 1D case so it would be nice to add it there but at least the 1D integrator's code is simple enough that a reader can quickly work it out.

Mateasek replied:
I added a line explaining the algorithm and expanded the param/ivar lines to give better explanation.

jacklovell commented on cherab/core/math/integrators/integrators2d.pyx:
```diff
@@ -59,3 +65,305 @@ cdef class Integrator2D:
+        # cache roots and weights for x dimension
+        n = (self._x_max_order + self._x_min_order) * (self._x_max_order - self._x_min_order + 1) // 2
+
+        self._x_roots = np.zeros(n, dtype=np.float64)
```
These are essentially 1D views of ragged arrays, right? Where each row would be the coefficients for a given order. They would be lower triangular matrices if `min_x_order` = 1. Maybe put a comment somewhere here mentioning this, as it'll save someone at a later date wasting time trying to work out why you're manually indexing instead of just using 2D - or 3D with [x, y, order] - arrays here (like I might have just done...).

Mateasek replied:
Here I went with what @vsnever originally designed for the 1D case and I like this approach. A 2D ragged array would be storing zeros and you would have to track which indexes are valid for the current order anyway. I'm not saying using this flattened case is hugely faster or more efficient, but I think it is as clean as using a 2D matrix for storing the data.

I added explanatory comments for both 1D and 2D cases

jacklovell commented on cherab/core/math/integrators/integrators2d.pyx:
```diff
@@ -59,3 +65,305 @@ cdef class Integrator2D:
+        self._y_roots_mv = self._y_roots
+        self._y_weights_mv = self._y_weights
+    
+    cpdef double profile_evaluate(self, int n, double x_lower, double x_upper, Function1D y_lower, Function1D y_upper):
```
What's the purpose of this function, and `evaluate_overhead` below? They're not called anywhere: are they left over from testing?

Mateasek replied:
I used it to test performance, but I forgot to remove it. Removed...

jacklovell commented on cherab/core/math/integrators/integrators2d.pyx:
```diff
@@ -59,3 +65,305 @@ cdef class Integrator2D:
+            int x_ibegin, y_ibegin
+            int i, j
+            double current_integral, previous_integral, y_contribution, error
+            double x, y, x_offset, y_offset, x_slope, y_slope, scale, rtol, y_lower_val, y_upper_val
```
rtol and scale aren't used

Mateasek replied:
The whole evaluate method was refactored and simplified. This should be fixed.

jacklovell commented on cherab/core/math/tests/test_integrators.py:
```diff
@@ -67,14 +72,119 @@ def test_properties(self):
+        self.assertEqual(quadrature.y_max_order, y_max_order)
+        self.assertEqual(quadrature.integrand, Arg2D("x"))
+
+    def test_integrate(self):
```
Add tests for `x_max_order  > y_max_order` and `x_max_order < y_max_order` to complement the existing `x_max_order == x_min_order`. This may help avoid future regressions.

Mateasek replied:
The test now iterates over max_order product of [30, 40, 50] and min_order product [10, 20, 30].

jacklovell reviewed (changes requested):
Very nice, only minor comments from me. Some of the (code comment related) comments apply equally to the existing `GaussianQuadrature` class on which this is based, so if you can add some clarifying code comments there too that would be nice but not a deal breaker.

Mateasek committed:
Merge branch 'development' into feature/GaussianQuadrature2D

Mateasek committed:
Rename GaussianQuadrature to GaussianQuadrature1D

To keep consistent naming with function frameword and quadrature2D
integrator.

GaussianQuadrature class marked as deprecated.

Mateasek committed:
remove cython numpy ndarray type and import

Mateasek committed:
Add comment explanation to root, weight storage

Mateasek committed:
Make docstring a raw string

Mateasek committed:
Improve docstring of GaussianQuadrature1D

Mateasek committed:
Refactor evaluate method, improve docstring

Docstring was improved, explanation of order increase added.
The evaluate method was refactored to improve efficiency

Redundant methods for testing were removed

Mateasek committed:
Improve tests

Multiple values and combinations of max_orders and min_orders are now
tested.

Mateasek committed:
Merge branch 'development' into feature/GaussianQuadrature2D

Mateasek committed:
Add GaussianQuadrature rename to changelog

jacklovell commented on docs/source/math/integrators.rst:
```diff
@@ -0,0 +1,11 @@
+Integrators
+-------------
+
+.. autoclass:: cherab.core.math.integrators.integrators1d.GaussianQuadrature
```
Should be `GaussianQuadrature1D`

Mateasek replied:
Good catch, I found few more `GaussianQuadrature` imports/usage and replaced it. I hope that it is all now.

Mateasek committed:
Replace deprecated GausianQuadrature usage

with GausianQuadrature1D

Mateasek committed:
Merge branch 'development' into feature/GaussianQuadrature2D

Mateasek committed:
Fix doctring formatting

Mateasek committed:
Merge branch 'development' into feature/GaussianQuadrature2D
